### PR TITLE
Ensure admin never sets task resource request > limit

### PR DESCRIPTION
**Pull request by katrogan**
# TL;DR
In the case where a user specifies a task resource limit but not the request value, admin will substitute the request value with the platform-configured default. In some cases this default can exceed the user-specified limit which is non-sensical and prevents kubernetes from ever scheduling the task.

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
Fixes user-reported bug.

## Tracking Issue
https://github.com/lyft/flyte/issues/264

## Follow-up issue
_NA_

### CODE REVIEW

File: pkg/manager/impl/execution_manager.go [hunk 1]
```diff
@@ -263,6 +265,41 @@ func assignResourcesIfUnset(ctx context.Context, identifier *core.Identifier,
 	return resourceEntries
 }
 
+func resolveTaskLimitsAndPlatformRequestDefaults(ctx context.Context, identifier *core.Identifier,
```
Comment (EngHabu) [inline pkg/manager/impl/execution_manager.go:268]: I don't see the code updating the limits (as the name suggests?) am I missing something?

Comment (katrogan) [reply pkg/manager/impl/execution_manager.go:268]: nope, I just struggled with naming. is 'considerTaskLimitsAndPlatformRequestDefaults` or `reconcileTaskLimitsAndPlatformRequestDefaults` any better? other suggestions welcomed

Comment (wild-endeavor) [reply pkg/manager/impl/execution_manager.go:268]: the limit is not supposed to change though... the request is.

**Review by wild-endeavor — APPROVED** [review]
just add a comment to the function to clear up confusion?

File: pkg/manager/impl/execution_manager.go (hunk 1 above)
Comment (katrogan) [reply pkg/manager/impl/execution_manager.go]: okay updated

**Comment by katrogan** [comment]
friendly ping @EngHabu @wild-endeavor @bnsblue

File: pkg/manager/impl/execution_manager.go [hunk 2]
```diff
@@ -263,6 +265,41 @@ func assignResourcesIfUnset(ctx context.Context, identifier *core.Identifier,
...
+		limitValue, ok := resourceLimits[resourceEntry.Name]
+		if !ok {
+			// Unexpected - at this stage both requests and limits should be populated.
+			logger.Warningf(ctx, "No limit specified for [%v] resource [%s] although request was set", identifier,
+				resourceEntry.Name)
+			continue
+		}
+		if quantity.Cmp(resource.MustParse(limitValue)) == 1 {
+			// The quantity is greater than the limit! Course correct below.
+			logger.Debugf(ctx, "Updating requested value for task [%+v] resource [%s]. Overriding to [%s] from [%s]",
```
Comment (bnsblue) [inline pkg/manager/impl/execution_manager.go:291]: nit: can you add what causes this override to the log so that it's clearer when reading the log?

Comment (bnsblue) [reply pkg/manager/impl/execution_manager.go:291]: Also I feel the debug level should be warning/info? Not quite sure 🤔

File: pkg/manager/impl/execution_manager_test.go [hunk 3]
```diff
@@ -2464,6 +2464,111 @@ func TestAssignResourcesIfUnset(t *testing.T) {
...
+					Name:  core.Resources_CPU,
+					Value: "2",
+				},
+				{
+					Name:  core.Resources_MEMORY,
+					Value: "1.5",
+				},
+			},
+		}, resources))
+	})
```
Comment (bnsblue) [inline pkg/manager/impl/execution_manager_test.go:2569]: Can you add a test that tests this logic? Specifically, can you add a test where one of the resource does not have the limit specified? 
https://github.com/lyft/flyteadmin/pull/126/files#diff-fc047e54b9dd82ca7c89ac9b32cb07b3R282-R288

**Comment by bnsblue** [comment]
LGTM. Just some nits.

File: pkg/manager/impl/execution_manager_test.go (hunk 3 above)
Comment (katrogan) [reply pkg/manager/impl/execution_manager_test.go:2569]: at the stage where the method is called all values will have already been substituted with the application defaults. this is why my test only verifies that if the defaults are weirdly configured we choose the min of acceptable values

Comment (bnsblue) [reply pkg/manager/impl/execution_manager_test.go:2569]: oh interesting. So the control flow should never enter this block?
https://github.com/lyft/flyteadmin/pull/126/files#diff-fc047e54b9dd82ca7c89ac9b32cb07b3R283-R288

Comment (katrogan) [reply pkg/manager/impl/execution_manager_test.go:2569]: unless the user is specifying a resource we don't provide defaults for. but in that case we don't find ourselves in the original conundrum where the substituted request > limit since there will be no limit default value which is a permissible situation for scheduling

File: pkg/manager/impl/execution_manager.go (hunk 2 above)
Comment (katrogan) [reply pkg/manager/impl/execution_manager.go]: updated & upgraded

File: pkg/manager/impl/execution_manager_test.go (hunk 3 above)
Comment (bnsblue) [reply pkg/manager/impl/execution_manager_test.go:2569]: From a unit-test point of view I think that it would still be nice to have a test for it since we don't know if in the future this function would be invoked through other code paths where the default limits were not pre-filled. But as you said it is not critical here.

**Review by bnsblue — APPROVED** [review]
